# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/MediaDecoder.h
# Commit: c57c41e8c39e
# Full Hash: c57c41e8c39ea51e96c13af8ecdcbe1640e2a9d6
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2019-11-14 17:12:31
# Regressor Bug: 1172394
# File Overlap Count: 1
# Description:
#   Bug 1172394 - Refactor how DecodedStream is set up. r=padenot
#   
#   This patch removes the responsibility of js-facing MediaStreamTracks from the
#   MediaDecoder stack, and moves the machinery for setting up DecodedStream to
#   higher order functions like state mirroring and watchables.
# ==============================================================================

diff -r a796541fe5ef -r c57c41e8c39e dom/media/MediaDecoder.h
--- a/dom/media/MediaDecoder.h	Wed Nov 13 08:55:29 2019 +0000
+++ b/dom/media/MediaDecoder.h	Wed Nov 13 08:55:39 2019 +0000
@@ -49,7 +49,6 @@
 class MediaFormatReader;
 class MediaDecoderStateMachine;
 struct MediaPlaybackEvent;
-struct SharedDummyTrack;
 
 enum class Visibility : uint8_t;
 
@@ -168,11 +167,11 @@
   // not connected to streams created by captureStreamUntilEnded.
 
   // Turn output capturing of this decoder on or off. If it is on, the
-  // MediaDecoderStateMachine will only create a MediaSink after output tracks
-  // have been set. This is to ensure that it doesn't create a regular MediaSink
+  // MediaDecoderStateMachine's media sink will only play after output tracks
+  // have been set. This is to ensure that it doesn't skip over any data
   // while the owner has intended to capture the full output, thus missing to
   // capture some of it. The owner of the MediaDecoder is responsible for adding
-  // output tracks while the output is captured.
+  // output tracks in a timely fashion while the output is captured.
   void SetOutputCaptured(bool aCaptured);
   // Add an output track. All decoder output for the track's media type will be
   // sent to the track.
@@ -615,6 +614,16 @@
 
   Canonical<bool> mLooping;
 
+  // Whether this MediaDecoder's output is captured. When captured, all decoded
+  // data must be played out through mOutputTracks.
+  Canonical<bool> mOutputCaptured;
+
+  // Tracks that, if set, will get data routed through them.
+  Canonical<nsTArray<RefPtr<ProcessedMediaTrack>>> mOutputTracks;
+
+  // PrincipalHandle to be used when feeding data into mOutputTracks.
+  Canonical<PrincipalHandle> mOutputPrincipal;
+
   // Media duration set explicitly by JS. At present, this is only ever present
   // for MSE.
   Maybe<double> mExplicitDuration;
@@ -647,6 +656,16 @@
     return &mPreservesPitch;
   }
   AbstractCanonical<bool>* CanonicalLooping() { return &mLooping; }
+  AbstractCanonical<bool>* CanonicalOutputCaptured() {
+    return &mOutputCaptured;
+  }
+  AbstractCanonical<nsTArray<RefPtr<ProcessedMediaTrack>>>*
+  CanonicalOutputTracks() {
+    return &mOutputTracks;
+  }
+  AbstractCanonical<PrincipalHandle>* CanonicalOutputPrincipal() {
+    return &mOutputPrincipal;
+  }
   AbstractCanonical<PlayState>* CanonicalPlayState() { return &mPlayState; }
 
  private: